halo2_gadgets/src/sinsemilla/chip.rs

```diff
@@ -186,77 +203,46 @@ where
-            config.double_and_add.Y_A(meta, rotation)
-        };
-
         // Check that the initial x_A, x_P, lambda_1, lambda_2 are consistent with y_Q.
         // https://p.z.cash/halo2-0.1:sinsemilla-constraints?partial
         meta.create_gate("Initial y_Q", |meta| {
             let q_s4 = meta.query_selector(config.q_sinsemilla4);
             let y_q = meta.query_fixed(config.fixed_y_q, Rotation::cur());
 
-            // Y_A = (lambda_1 + lambda_2) * (x_a - x_r)
-            let Y_A_cur = Y_A(meta, Rotation::cur());
+            // y_a_derived = (lambda_1 + lambda_2) * (x_a - x_r)
+            let y_a_derived = config.double_and_add.y_a(meta, Rotation::cur());
 
-            // 2 * y_q - Y_{A,0} = 0
-            let init_y_q_check = y_q * two - Y_A_cur;
-
-            Constraints::with_selector(q_s4, Some(("init_y_q_check", init_y_q_check)))
+            Constraints::with_selector(q_s4, Some(("init_y_q_check", y_q - y_a_derived)))
         });
 
-        // https://p.z.cash/halo2-0.1:sinsemilla-constraints?partial
-        meta.create_gate("Sinsemilla gate", |meta| {
-            let q_s1 = meta.query_selector(config.q_sinsemilla1);
-            let q_s3 = config.q_s3(meta);
-
-            let lambda_1_next = meta.query_advice(config.double_and_add.lambda_1, Rotation::next());
-            let lambda_2_cur = meta.query_advice(config.double_and_add.lambda_2, Rotation::cur());
+        // Final double-and-add gate
+        meta.create_gate("final check", |meta| {
+            // x_{A,i}
             let x_a_cur = meta.query_advice(config.double_and_add.x_a, Rotation::cur());
+            // x_{A,i-1}
             let x_a_next = meta.query_advice(config.double_and_add.x_a, Rotation::next());
+            // λ_{2,i}
+            let lambda2_cur = meta.query_advice(config.double_and_add.lambda_2, Rotation::cur());
+            let y_a_cur = config.double_and_add.y_a(meta, Rotation::cur());
 
-            // x_r = lambda_1^2 - x_a_cur - x_p
-            let x_r = x_r(meta, Rotation::cur());
+            // Final round selector: q_s1 * (q_s3 / 2)
+            let selector = {
+                let q_s1 = meta.query_selector(q_sinsemilla1);
+                let q_s3_div_2 = q_s3(meta, q_sinsemilla2) * pallas::Base::TWO_INV;
 
-            // Y_A = (lambda_1 + lambda_2) * (x_a - x_r)
-            let Y_A_cur = Y_A(meta, Rotation::cur());
-
-            // Y_A = (lambda_1 + lambda_2) * (x_a - x_r)
-            let Y_A_next = Y_A(meta, Rotation::next());
-
-            // lambda2^2 - (x_a_next + x_r + x_a_cur) = 0
-            let secant_line =
-                lambda_2_cur.clone().square() - (x_a_next.clone() + x_r + x_a_cur.clone());
-
-            // lhs - rhs = 0, where
-            //    - lhs = 4 * lambda_2_cur * (x_a_cur - x_a_next)
-            //    - rhs = (2 * Y_A_cur + (2 - q_s3) * Y_A_next + 2 * q_s3 * y_a_final)
-            let y_check = {
-                // lhs = 4 * lambda_2_cur * (x_a_cur - x_a_next)
```

Comment:
To check equivalence with `v0.2.0` Sinsemilla (used in the Orchard circuit), we could allow the configuration to enable legacy behaviour somehow.